python2-cython/challenge/basics.pyx

```diff
@@ -0,0 +1,252 @@
+"""Definitions of Python extension types for basics module
+"""
+
+from libcpp cimport bool
+from libcpp.memory cimport shared_ptr
+from cython.operator cimport dereference as deref
+from cpython.object cimport Py_EQ, Py_NE
+
+from _basics cimport move
+from _basics cimport compare as _compare
+from _basics cimport adjacent as _adjacent
+from basics cimport Doodad
+from _basics cimport Doodad as _Doodad
+from _basics cimport Secret as _Secret
+from _basics cimport WhatsIt as _WhatsIt
+
+cpdef compare(Secret a, Secret b):
+    return _compare(deref(a.thisptr), deref(b.thisptr))
+
+cpdef adjacent(Secret a, Secret b):
+    return _adjacent(deref(a.thisptr), deref(b.thisptr))
+
+cdef _WhatsIt from_tuple(tuple t):
+    """Helper function only visible from Cython.
+    """
+
+    cdef _WhatsIt w = _WhatsIt(t[0], t[1])
+
+    return w
+
+cdef class Secret:
+    """Opaque type only for use by Doodad.
+    """
+
+    cdef const _Secret *thisptr
+
+cdef class Doodad:
+    """Python interface to C++ type Doodad.
+
+    Parameters
+    ----------
+    name : str | tuple
+        When tuple, a pair of name, value. Otherwise a string.
+    value : int
+        A value
+    """
+
+    def __init__(self, name=None, value=1, init=True):
+        if init:
+            if isinstance(name, tuple):
+                self.thisptr.reset(new _Doodad(from_tuple(name)))
+            elif isinstance(name, str):
+                self.thisptr.reset(new _Doodad(name, value))
+            else:
+                raise TypeError("a string or tuple is required")
```

Comment:
Could you remove the need for the `init` argument somehow?  Seems like an implementation detail the user shouldn't see.  Maybe by implementing `__new__`, and having that take a `thisptr` argument?